plugins/gstperf.c

```diff
@@ -198,13 +243,89 @@ gst_perf_get_property (GObject * object, guint property_id,
   }
 }
 
+static gboolean
+gst_perf_update_bps (void *data)
+{
+  guint buffer_current_idx;
+  GstPerf *perf;
+  guint byte_count;
+  gdouble bps, mean_bps;
+
+  g_return_val_if_fail (data, FALSE);
+
+  perf = GST_PERF (data);
+
+  g_mutex_lock (&perf->byte_count_mutex);
+  byte_count = perf->byte_count;
+  perf->byte_count = G_GUINT64_CONSTANT (0);
+  g_mutex_unlock (&perf->byte_count_mutex);
+
+  g_mutex_lock (&perf->mean_bps_mutex);
+  mean_bps = perf->mean_bps;
+  g_mutex_unlock (&perf->mean_bps_mutex);
+
+  /* Calculate bits per second */
+  bps =
+      byte_count * GST_PERF_BITS_PER_BYTE / (perf->bps_running_interval /
+      GST_PERF_MS_PER_S);
+
+  /* Update bps average */
+  if (!perf->bps_window_size) {
+    mean_bps = gst_perf_update_average (perf->byte_count_total, bps, mean_bps);
+  } else {
+    /*
+     * Moving average uses a circular buffer, get index for next value which
+     * is the oldest sample, this is the same as the value were the new sample
+     * is to be stored
+     */
+    buffer_current_idx = (perf->byte_count_total) % perf->bps_window_size;
+
+    mean_bps =
+        gst_perf_update_moving_average (perf->bps_window_size, mean_bps,
+        bps, perf->bps_window_buffer[buffer_current_idx]);
+
+    perf->bps_window_buffer[buffer_current_idx] = bps;
+  }
+  g_mutex_lock (&perf->mean_bps_mutex);
+  perf->mean_bps = mean_bps;
+  g_mutex_unlock (&perf->mean_bps_mutex);
+
+  g_mutex_lock (&perf->bps_mutex);
+  perf->bps = bps;
+  g_mutex_unlock (&perf->bps_mutex);
+
+  perf->byte_count_total++;
+
+  g_signal_emit_by_name (perf, "on-bitrate", mean_bps);
+
+  return TRUE;
+}
+
 static gboolean
 gst_perf_start (GstBaseTransform * trans)
 {
   GstPerf *perf = GST_PERF (trans);
 
   gst_perf_clear (perf);
 
+  g_mutex_init (&perf->byte_count_mutex);
```

Comment:
Why only this mutex? I would just do it once in the instance `init` method.  and avoid init and clear.